### DISCUSSION TITLE
Text block renders outside of its parent

### DISCUSSION THREAD

**Discussion by YohDeadfall** [AvaloniaUI/Avalonia#13203] [Q&A]
For some reason `TextBlock` might be rendered outside of its parent as shown on the image below. Without `UniformGridLayout` it cannot be reproduced, with it it doesn't matter which orientation is used. Replacing `TextBlock` by `TextBox` makes rendering correct, but the height for items is limited by the first one.

![image](https://user-images.githubusercontent.com/10694013/107800306-eae28800-6d5e-11eb-9dd7-23c17b54a614.png)

## Minimal code to repro

```csharp
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;

namespace TextBlockBug
{
    public class MainWindow : Window
    {
        public MainWindow()
        {
            AvaloniaXamlLoader.Load(this);

            this.FindControl<ItemsRepeater>("Repeater")
                .Items = new[]
                {
                    new SomeViewModel("50mm"),
                    new SomeViewModel("CameraRoundtrip"),
                    new SomeViewModel("EffectSettings"),
                    new SomeViewModel("Material_Preview"),
                    new SomeViewModel("PBRTestSequence"),
                    new SomeViewModel("PostProcessingSettings"),
                    new SomeViewModel("Scenes"),
                };

            this.AttachDevTools();
        }
    }

    public record SomeViewModel(string Name);
}
```

```xml
<Window
    x:Class="TextBlockBug.MainWindow"
    xmlns="https://github.com/avaloniaui"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    Title="TextBlockBug"
    d:DesignHeight="450"
    d:DesignWidth="800"
    mc:Ignorable="d">
    <ScrollViewer HorizontalScrollBarVisibility="Disabled">
        <ItemsRepeater Name="Repeater">
            <ItemsRepeater.Layout>
                <UniformGridLayout
                    MinColumnSpacing="2"
                    MinRowSpacing="2"
                    Orientation="Horizontal" />
            </ItemsRepeater.Layout>
            <ItemsRepeater.ItemTemplate>
                <DataTemplate>
                    <Border
                        Margin="2"
                        Background="Green"
                        BorderBrush="Black"
                        BorderThickness="1">
                        <Grid RowDefinitions="Auto,Auto">
                            <Panel
                                Grid.Row="0"
                                Width="64"
                                Height="64"
                                Background="Green" />
                            <TextBlock
                                Grid.Row="1"
                                MaxWidth="64"
                                Padding="2"
                                HorizontalAlignment="Center"
                                Text="{Binding Name}"
                                TextAlignment="Center"
                                TextWrapping="Wrap" />
                        </Grid>
                    </Border>
                </DataTemplate>
            </ItemsRepeater.ItemTemplate>
        </ItemsRepeater>
    </ScrollViewer>
</Window>
```

**Comment by grokys** [comment]
I assigned @Gillibald and the unassigned him because i noticed "Without UniformGridLayout it cannot be reproduced", so probably not a text issue as such...

So looks like this might be a problem with `ItemsRepeater`.

**Comment by lxymahatma** [comment]
I think it might be the problem with `ItemsRepeater` and `Grid`'s "Auto" Row definition? (not sure but might be)
![image](https://github.com/AvaloniaUI/Avalonia/assets/66151783/03d39b4d-54e0-427b-a9a8-9af1a39ff4f1)
![image](https://github.com/AvaloniaUI/Avalonia/assets/66151783/ab7e5fb8-acaf-4294-b694-6fe03d1877e6)
i've met this scenario which the Download button is rendered out of the card
However, set the second row to a constant like 100 won't cause this problem
Also, use `StackPanelLayout` in `ItemsRepeater` doesn't cause this problem
Using `ListBox` with `UniformGrid` isn't causing this problem as well

So i guess might be some conflicts with `ItemsRepeater` and "Auto" Row definition property in Grid

**Comment by maxkatz6** [comment]
UniformGridLayout estimates height depending on the first item, IIRC. In combination with this Grid layout, it does seem expected for the repeater.
If you want to avoid rendering outside, add ClipToBounds=true on the parent. Or add a scroll viewer, or a popup with detailed content.

ListBox with UniformGrid won't have this problem, as it measures all children before arranging them, as it's not virtualized.

**Comment by lxymahatma** [comment]
> UniformGridLayout estimates height depending on the first item, IIRC. In combination with this Grid layout, it does seem expected for the repeater. If you want to avoid rendering outside, add ClipToBounds=true on the parent. Or add a scroll viewer, or a popup with detailed content.
> 
> ListBox with UniformGrid won't have this problem, as it measures all children before arranging them, as it's not virtualized.

`ClipToBounds=True` doesn't seem to solve my problem (but maybe cz I'm pretty new to avalonia and have never used this so I use it in the wrong way)
Anyway thanks for the clear explanation 👍
In this case I think yeah it's the expected result if it only evaluated the first item height
![image](https://github.com/AvaloniaUI/Avalonia/assets/66151783/92cc98cb-973c-4574-8139-cc5ce47d8027)
This is my code after adding `ClipToBounds=True`, the download button still not showing in the card

**Comment by timunie** [comment]
The thing is, that the Grid can be bigger than the Card actually is. So it will clip (means cut) the bottom. I'd use GridSize `*` in the second row, so the title always has the same height. Then use TextTrimming to show "..." is text is too long.

**Comment by lxymahatma** [comment]
> The thing is, that the Grid can be bigger than the Card actually is. So it will clip (means cut) the bottom. I'd use GridSize `*` in the second row, so the title always has the same height. Then use TextTrimming to show "..." is text is too long.

i've done TextTrimming with third row, which shows the author of the map, but the second row is the name of the map so it might be better to show the full name instead
I'm considering redesign it with stackpanel but idk whether it will solve the issue or not (but if it only estimate size for the first item i don't think it would work 😂)

Edit: I feel like most people don't want to read long names as well, they just use abbreviations so i think i'm just gonna use your method lol

**Comment by timunie** [comment]
in ItemsRepeater all items have same size. So you have some options:
- Use abbrevations
- Increase Card size
- Add a ScrollViewer to your Card 
- Use another Layout (ItemsControl)

I still see that as by design, so will convert this issue into a discussion now. I hope you understand this.
